src/gov-action-contracts/AIPs/AIP4844/AIP4844Action.sol

```diff
@@ -0,0 +1,114 @@
+        address _newSequencerInboxImpl,
+        address _newChallengeMangerImpl,
+        ProxyAdmin _govProxyAdmin,
+        address _newOsp
+    ) {
+        require(
+            Address.isContract(address(_l1AddressRegistry)),
+            "AIP4844Action: _11AddressRegistry is not a contract"
+        );
+        l1AddressRegistry = _l1AddressRegistry;
+
+        require(_newWasmModuleRoot != bytes32(0), "AIP4844Action: _newWasmModuleRoot is empty");
+        newWasmModuleRoot = _newWasmModuleRoot;
+
+        require(
+            Address.isContract(_newSequencerInboxImpl),
+            "AIP4844Action: _newSequencerInboxImpl is not a contract"
+        );
+        newSequencerInboxImpl = _newSequencerInboxImpl;
+
+        require(
+            Address.isContract(_newChallengeMangerImpl),
+            "AIP4844Action: _newChallengeMangerImpl is not a contract"
+        );
+        newChallengeManagerImpl = _newChallengeMangerImpl;
+
+        require(
+            Address.isContract(address(_govProxyAdmin)),
+            "AIP4844Action: _govProxyAdmin is not a contract"
+        );
+        govProxyAdmin = _govProxyAdmin;
+
+        require(Address.isContract(address(_newOsp)), "AIP4844Action: _newOsp is not a contract");
+        newOsp = _newOsp;
+    }
+
+    function perform() public {
+        IRollupCore rollup = l1AddressRegistry.rollup();
+        IRollupAdmin(address(rollup)).setWasmModuleRoot(newWasmModuleRoot);
+
+        // verify:
+        require(
+            rollup.wasmModuleRoot() == newWasmModuleRoot, "AIP4844Action: wasm module root not set"
+        );
+
+        TransparentUpgradeableProxy sequencerInbox =
+            TransparentUpgradeableProxy(payable(address(l1AddressRegistry.sequencerInbox())));
+        (, uint256 futureBlocksBefore,,) =
+            ISequencerInbox(address(sequencerInbox)).maxTimeVariation();
+        govProxyAdmin.upgradeAndCall(
+            sequencerInbox,
+            newSequencerInboxImpl,
+            abi.encodeWithSelector(ISeqInboxPostUpgradeInit.postUpgradeInit.selector)
+        );
+
+        // verify
+        require(
+            govProxyAdmin.getProxyImplementation(sequencerInbox) == newSequencerInboxImpl,
+            "AIP4844Action: new seq inbox implementation set"
+        );
+        (, uint256 futureBlocksAfter,,) =
+            ISequencerInbox(address(sequencerInbox)).maxTimeVariation();
+        require(
+            futureBlocksBefore != 0 && futureBlocksBefore == futureBlocksAfter,
+            "AIP4844Action: maxTimeVariation not set"
+        );
+
+        // set the new challenge manager impl
+        TransparentUpgradeableProxy challengeManager = TransparentUpgradeableProxy(
+            payable(address(l1AddressRegistry.rollup().challengeManager()))
+        );
+        govProxyAdmin.upgradeAndCall(
+            challengeManager,
+            newChallengeManagerImpl,
+            abi.encodeWithSelector(IChallengeManagerUpgradeInit.postUpgradeInit.selector, newOsp)
```

Comment:
usually prefer to use encodeCall for type check

Reply:
updated